tsgo loses all type narrowing in a `switch (typeof x)` when any clause is `case ""`

## Steps to reproduce

```ts
function f(x: string | number) {
  switch (typeof x) {
    case "":
    case "string":
      x.charAt(0);
      break;
  }
}
```

## Behavior with `typescript@6.0`

The type of `x` is correctly narrowed to `string` and the typechecker correctly produces one error:

```
bug.ts:3:10 - error TS2678: Type '""' is not comparable to type '"string" | "number" | "bigint" | "boolean" | "symbol" | "undefined" | "object" | "function"'.

3     case "":
           ~~
```

## Behavior with `tsgo`

The type of `x` is not narrowed correctly and the typechecker incorrectly produces an additional error:

```
bug.ts:3:10 - error TS2678: Type '""' is not comparable to type '"bigint" | "boolean" | "function" | "number" | "object" | "string" | "symbol" | "undefined"'.

3     case "":
           ~~

bug.ts:5:9 - error TS2339: Property 'charAt' does not exist on type 'string | number'.
  Property 'charAt' does not exist on type 'number'.

5       x.charAt(0);
          ~~~~~~
```

--- a/internal/checker/flow.go
+++ b/internal/checker/flow.go
@@ -1970,15 +1970,11 @@ func (c *Checker) getSwitchClauseTypeOfWitnesses(node *ast.Node) []string {
 		witnesses := make([]string, len(clauses))
 		for i, clause := range clauses {
 			if clause.Kind == ast.KindCaseClause {
-				var text string
-				if ast.IsStringLiteralLike(clause.Expression()) {
-					text = clause.Expression().Text()
-				}
-				if text == "" {
+				if !ast.IsStringLiteralLike(clause.Expression()) {
 					witnesses = nil
 					break
 				}
-				if !slices.Contains(witnesses, text) {
+				if text := clause.Expression().Text(); !slices.Contains(witnesses, text) {
 					witnesses[i] = text
 				}
 			}

--- a/testdata/baselines/reference/compiler/typeofSwitchEmptyStringCase.errors.txt
+++ b/testdata/baselines/reference/compiler/typeofSwitchEmptyStringCase.errors.txt
@@ -1,0 +1,17 @@
+typeofSwitchEmptyStringCase.ts(5,10): error TS2678: Type '""' is not comparable to type '"bigint" | "boolean" | "function" | "number" | "object" | "string" | "symbol" | "undefined"'.
+
+
+==== typeofSwitchEmptyStringCase.ts (1 errors) ====
+    // https://github.com/microsoft/typescript-go/issues/3909
+    
+    function f(x: string | number) {
+      switch (typeof x) {
+        case "":
+             ~~
+!!! error TS2678: Type '""' is not comparable to type '"bigint" | "boolean" | "function" | "number" | "object" | "string" | "symbol" | "undefined"'.
+        case "string":
+          x.charAt(0);
+          break;
+      }
+    }
+    

--- a/testdata/baselines/reference/compiler/typeofSwitchEmptyStringCase.symbols
+++ b/testdata/baselines/reference/compiler/typeofSwitchEmptyStringCase.symbols
@@ -1,0 +1,23 @@
+//// [tests/cases/compiler/typeofSwitchEmptyStringCase.ts] ////
+
+=== typeofSwitchEmptyStringCase.ts ===
+// https://github.com/microsoft/typescript-go/issues/3909
+
+function f(x: string | number) {
+>f : Symbol(f, Decl(typeofSwitchEmptyStringCase.ts, 0, 0))
+>x : Symbol(x, Decl(typeofSwitchEmptyStringCase.ts, 2, 11))
+
+  switch (typeof x) {
+>x : Symbol(x, Decl(typeofSwitchEmptyStringCase.ts, 2, 11))
+
+    case "":
+    case "string":
+      x.charAt(0);
+>x.charAt : Symbol(String.charAt, Decl(lib.es5.d.ts, --, --))
+>x : Symbol(x, Decl(typeofSwitchEmptyStringCase.ts, 2, 11))
+>charAt : Symbol(String.charAt, Decl(lib.es5.d.ts, --, --))
+
+      break;
+  }
+}
+

--- a/testdata/baselines/reference/compiler/typeofSwitchEmptyStringCase.types
+++ b/testdata/baselines/reference/compiler/typeofSwitchEmptyStringCase.types
@@ -1,0 +1,30 @@
+//// [tests/cases/compiler/typeofSwitchEmptyStringCase.ts] ////
+
+=== typeofSwitchEmptyStringCase.ts ===
+// https://github.com/microsoft/typescript-go/issues/3909
+
+function f(x: string | number) {
+>f : (x: string | number) => void
+>x : string | number
+
+  switch (typeof x) {
+>typeof x : "bigint" | "boolean" | "function" | "number" | "object" | "string" | "symbol" | "undefined"
+>x : string | number
+
+    case "":
+>"" : ""
+
+    case "string":
+>"string" : "string"
+
+      x.charAt(0);
+>x.charAt(0) : string
+>x.charAt : (pos: number) => string
+>x : string
+>charAt : (pos: number) => string
+>0 : 0
+
+      break;
+  }
+}
+

--- a/testdata/tests/cases/compiler/typeofSwitchEmptyStringCase.ts
+++ b/testdata/tests/cases/compiler/typeofSwitchEmptyStringCase.ts
@@ -1,0 +1,12 @@
+// @noEmit: true
+
+// https://github.com/microsoft/typescript-go/issues/3909
+
+function f(x: string | number) {
+  switch (typeof x) {
+    case "":
+    case "string":
+      x.charAt(0);
+      break;
+  }
+}

